### ISSUE TITLE
Allow other Result types in `#[pyfunction]` etc

### ISSUE THREAD

**Issue by davidhewitt** [PyO3/pyo3#1106]
At the moment the only `Result` type we allow in `#[pyfunction]` is `PyResult`. 

After a question on Gitter yesterday I was thinking if we can allow instead any `Result` which has `E: Into<PyErr>`.

That would allow code like this (all implementations replaced with `...`):

```
enum MyError { ... }

impl From<MyError> for PyErr { ... }

#[pyfunction]
fn my_err_function() -> Result<(), MyErr> { ... }
```

Pros:
- Extra flexibility for users implementing `#[pyfunction]`
- Error handling improvement when calling this `#[pyfunction]` in Rust - can handle `MyError` typed enum rather than dynamic `PyErr`.
- Possible performance improvement when calling this `#[pyfunction]` in Rust - no need to create `PyErr` which is relatively higher-cost than `MyError`.

Cons:
- Making this accept more types might over-complicate pyo3's API. I'm not sure it's so bad, though.

cc @marioortizmanero who posted on this topic on Gitter. (At least, I think that was what was being asked!)

**Comment by davidhewitt** [comment]
Broadly speaking, I don't think the implementation is too hard. I think we can just replace these lines in `callback.rs`:

https://github.com/PyO3/pyo3/blob/4840619d63e09fc885970a33057cb02b68969403/src/callback.rs#L40-L47

with this implementation:

```
impl<T, E, U> IntoPyCallbackOutput<U> for PyResult<T, E>
where
    T: IntoPyCallbackOutput<U>,
    E: Into<PyErr>
{
    fn convert(self, py: Python) -> PyResult<U> {
        self.map_err(Into::into).and_then(|t| t.convert(py))
    }
}
```

However, at the moment this causes `pyo3` to fail to compile because type inference for the new parameter `E` is not possible. I think this type inference failure only applies to some internal implementations of our `#[pyproto]`, and for external code this would not be a problem.

So our internal code would need a little refactoring, but otherwise it's reasonably straightforward.

**Comment by birkenfeld** [comment]
Note that the question mark operator already calls into() so I don't think the flexibility implementing the pyfunction is really needed. The other pros might be nice though.

**Comment by marioortizmanero** [comment]
Yes, I'm the one who suggested this feature, and you got it right.

I think that as long as it's well documented, it shouldn't be an issue. This should be mentioned in the [type conversions page of the book](https://pyo3.rs/v0.11.1/conversions.html) with a note that the wrapped Error must implement `E: Into<PyErr>`.

**Comment by davidhewitt** [comment]
@marioortizmanero if you're interested in implementing this then you're very welcome to provide a PR! I'm happy to review or answer any questions you have.

**Comment by marioortizmanero** [comment]
Thanks @davidhewitt, I've opened a new PR with the proposed changes. Please take a look and let me know.

**Comment by marioortizmanero** [comment]
This can be closed now as #1118 was merged, @davidhewitt

**Comment by davidhewitt** [comment]
Thanks 👍
